allennlp/interpret/influence_interpreters/influence_interpreter.py

```diff
@@ -0,0 +1,111 @@
+from typing import List, Optional
+
+import torch
+
+from allennlp.common import Registrable
+from allennlp.common.util import JsonDict
+from allennlp.models.model import Model
+from allennlp.predictors import Predictor
+from allennlp.data.dataset_readers import DatasetReader
+from allennlp.data.data_loaders import MultiProcessDataLoader
+
+
+class InfluenceInterpreter(Registrable):
+    """
+    A `SaliencyInterpreter` interprets an AllenNLP Predictor's outputs by assigning an influence
+    score to each training instance with respect to each input.
+    """
+
+    def __init__(
+        self,
+        predictor: Predictor,
+        train_filepath: str,
+        test_filepath: str,
+        train_dataset_reader: DatasetReader,
+        test_dataset_reader: Optional[DatasetReader] = None,
+        params_to_freeze: List[str] = None,
+        k: int = 20,
+        device: int = -1,
+    ) -> None:
+        if type(device) is not int:
+            raise ValueError("'device' has to be int, -1 for cpu")
+        if type(device) is not int:
+            raise ValueError("'k' (i.e. number of supporting examples) has to be int")
+
+        self.predictor = predictor
+        self.model: Model = self.predictor._model
+        self.vocab = self.model.vocab
+        self.train_dataset_reader = train_dataset_reader
+        self.test_dataset_reader = test_dataset_reader or train_dataset_reader
+
+        self._device = torch.device(
+            f"cuda:{int(device)}" if torch.cuda.is_available() and not device >= 0 else "cpu"
+        )
+        # Dataloaders for going through train/test set (1 by 1)
+        self._train_loader = MultiProcessDataLoader(
+            self.train_dataset_reader, train_filepath, batch_size=1
+        )
+        self._train_loader.set_target_device(self._device)
+        self._train_loader.index_with(self.vocab)
+        self._test_loader = MultiProcessDataLoader(
+            self.test_dataset_reader, test_filepath, batch_size=1
+        )
+        self._test_loader.set_target_device(self._device)
+        self._test_loader.index_with(self.vocab)
+
+        # Number of supporting training instances has to be lass than the size of train set
+        self._k = min(k, len(self._train_loader))
+
+        self.model.to(self._device)
+
+        # so far, we assume all parameters are tuned during training
+        # we use freeze, because when model is loaded from archive, the requires_grad
+        # flag will be re-initialized to be true
+        if params_to_freeze is not None:
+            self.freeze_model(self.model, params_to_freeze, verbose=True)
+        # self._used_name2params = None  # some parameters might not be used.
+        # this is not set until we actually run the calculation, because some parameters might not be used.
+        self._used_params = None
+        self._used_params_name = None
+
+    def influence_interpret_from_json(self, inputs: JsonDict) -> JsonDict:
```

Comment:
Why do we need this method if we pass in a `test_data_path` in `__init__`? And it seems that `SimpleInfluence` doesn't implement this method.

Reply:
No we don't. I started this method by mimicking the `attack_from_json` as in `Attackers`, but later deprecated. Again, forget to delete...